solutions/1000-1099/1000-minimum-cost-to-merge-stones-hard.md

```diff
@@ -0,0 +1,322 @@
+we can stop here: our total cost = 8 + 17 = 25
+
+## The first thing we need to check can we merge all these piles into 1 pile ?
+we can check this : ## if((n-1)%(k-1) == 0) then only we can do merging.
+
+let's see how we can arrive at this formula.
+
+if there's n piles given and k piles are to be merged. Then,
+After the first merge : the length of the array will be n - (k-1).
+After the second merge : the length of the array will be n - (k-1) - (k-1).
+After the third merge : the length of the array will be n - (k-1) - (k-1) - (k-1)
+and so on...
+
+for eg:
+[0 1 2 3 4] -> indexing
+{3, 5, 1, 2, 6} -> array elements
+length is 5
+
+if we do single merge of {5, 1, 2} => {8}
+
+[0 1 2] -> indexing
+{3, 8, 6} -> array elements
+length is 3 : which n - (k-1) = 5 - (k-1) = 5 - (3-1) = 5 - 2 = 3;
+
+So, we know that after m merges the length must be 1, if we can merge piles
+So, we can say that Cost for single merge is k-1 and for m merges will be m*(k-1);
+
+total_merges = m * (k-1);
+
+### Given in the question : A move consists of merging exactly k consecutive piles into one pile and in the end only single pile is left.,
+
+So, we can say that:
+N - total_merges = 1
+
+N - m*(k-1) = 1;
+N - 1 = m*(k-1);
+m = (N-1) / (k-1)
+
+now, "m" must be positive number then only we can merge them
+
+In, this example: {3, 5, 1, 2, 6} here N = 5, k = 3
+m = (N-1) / (k-1)
+m = (5-1) / (3-1) = (4) / (2) = 2
+
+2 is positive integer, so here we can merge 2 times, which is actually true in our case.
+
+Here we need range sum which will be added in the cost. 
+for this purpose we will create prefix_Sum and use it.
+
+To calculate cost : we will use prefixSum.
+for eg:
+
+[0 1 2 3 4] -> indexing
+{3, 5, 1, 2, 6} -> array elements
+
+prefix sum is
+[0 1 2 3 4 5] -> indexing
+{0, 3, 8, 9, 11, 17} -> array elements
+
+if we are merging let's say {5, 1, 2}, then i = 1, j = 3 index.
+then cost will be prefix[j+1] - prefix[i] = prefix[3+1] - prefix[1] = 11 - 3 = 8 which is true.
+
+Also, before calculating cost : for every range we will check, if we can merge this or not
+if((j-i)%(piles-1) == 0) here j - i = current range and piles is k.
+
+Also, in the induction step : In the for loop we will increment by k = k + piles - 1; because after merging, 
+k-1 piles will be deleted for sure (in this case 2 piles will be deleted after every merge), 
+so we just start from next valid (3rd in this case) pile from current.
+
+Sub Problem will range from : 
+if we rec(stones, i, k) + rec(stones, k+1, j) : as in our for loop we will loop till second last_index, 
+so that for rec(k+1, j) : here k+1 remains valid.
+
+Driver Function Call : return rec(stones, 0, n-1, k); : we will start from 0th index and last index n-1 which is the valid range..
+
+## Approach 1: Pure Recursion : MCM Pattern
+**TC: O((N*(2^N))/K) here as we increment by k piles in every step and for every element we have 2 options and there are 2^N ways** 
+**SC: O(N) for prefix sum + O(N) + auxilary stack space**
```

Comment:
SC -> Space Complexity. Use full form as some beginners may not know the short form.